### PR TITLE
[SV] Add pass to hide non-synthesizable ops

### LINKED ISSUE #10378: [Seq] FIFO lowering: emits unguarded `assert property`, breaking synthesis
This issue written by Claude Opus 4.7 then edited by me.

-----

# `seq.fifo` lowering emits unguarded `assert property`, breaking Quartus (and other) synthesis

## Summary

The `seq.fifo` → HW lowering in [`LowerSeqFIFO.cpp`](https://github.com/llvm/circt/blob/main/lib/Dialect/Seq/Transforms/LowerSeqFIFO.cpp#L181-L201) unconditionally inserts two concurrent assertions ("FIFO empty when read enabled" / "FIFO full when write enabled"). After [`VerifToSV`](https://github.com/llvm/circt/blob/main/lib/Conversion/VerifToSV/VerifToSV.cpp#L195-L205), these are emitted as bare SystemVerilog `assert property` statements with no `` `ifndef SYNTHESIS `` guard and no other vendor-pragma wrap.

This breaks Intel Quartus, which rejects concurrent assertions in synthesis:

```
Error (13224): Verilog HDL or VHDL error at ListVoidResultCountAdapter.sv(43):
  Found a non-synthesizable construct "temporal assertion".
  If this construct is not required for design elaboration it must be wrapped in
  `no_translate synthesis pragmas
Error (13224): ... Line: 44 ...
```

The same pattern is also problematic for several other synthesis tools that don't fully support SVA, even when they accept the rest of the file.

## Reproduction

Any design that uses `seq.fifo` (directly or transitively — e.g. PyCDE's `ListWindowToSerial`, ESI runtime adapters, etc.) produces output containing:

```sv
FIFO_empty_when_read_enabled:
  assert property (@(posedge clk) ~(load_count & fifo_empty));
FIFO_full_when_write_enabled:
  assert property (@(posedge clk) ~(_GEN & fifo_full));
```

These statements are emitted at module scope, outside any guard.

## Root cause

`LowerSeqFIFO` creates `verif::ClockedAssertOp` ops directly into the module body at [LowerSeqFIFO.cpp#L189-L194](https://github.com/llvm/circt/blob/main/lib/Dialect/Seq/Transforms/LowerSeqFIFO.cpp#L189-L194):

```cpp
verif::ClockedAssertOp::create(
    rewriter, loc, notEmptyAndRden, verif::ClockEdge::Pos, clkI1,
    /*enable=*/Value(),
    rewriter.getStringAttr("FIFO empty when read enabled"));
```

`VerifClockedAssertLikeConversion` in [VerifToSV.cpp#L195-L205](https://github.com/llvm/circt/blob/main/lib/Conversion/VerifToSV/VerifToSV.cpp#L195-L205) then turns each into an `sv::AssertPropertyOp`, again with no synthesis guard. ExportVerilog prints them verbatim.

By contrast, the FIRRTL → HW lowering carefully wraps verification statements with `addToIfDefBlock("SYNTHESIS", ...)` (see e.g. [LowerToHW.cpp#L5235](https://github.com/llvm/circt/blob/main/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L5235) and [FirRegLowering.cpp#L413](https://github.com/llvm/circt/blob/main/lib/Conversion/SeqToSV/FirRegLowering.cpp#L413)), producing `` `ifndef SYNTHESIS ... `endif `` blocks that all major synthesis tools accept by defining the `SYNTHESIS` macro.

The existing knobs do not help:
- `LoweringOptions::disallowClockedAssertions` ([LoweringOptions.h#L178](https://github.com/llvm/circt/blob/main/include/circt/Support/LoweringOptions.h#L178)) only converts the form (concurrent → procedural always block); the result is still a synthesizable-construct error in Quartus.
- `-sv-extract-test-code` was removed (see [Firtool.cpp#L814](https://github.com/llvm/circt/blob/main/lib/Firtool/Firtool.cpp#L814), comment "`-sv-extract-test-code` has been removed").
- There is no flag that simply drops or guards verification ops produced by non-FIRRTL frontends.

So any frontend that uses `seq.fifo` (PyCDE, hand-written MLIR, Calyx-derived flows, etc.) cannot produce Quartus-clean SystemVerilog without post-processing the output file.

#### 1. Wrap FIFO assertions in `` `ifndef SYNTHESIS `` in `LowerSeqFIFO`

Smallest, most contained change. In [LowerSeqFIFO.cpp#L181-L201](https://github.com/llvm/circt/blob/main/lib/Dialect/Seq/Transforms/LowerSeqFIFO.cpp#L181-L201), wrap the two `verif::ClockedAssertOp::create(...)` calls in an `sv::IfDefOp("SYNTHESIS", /*then*/ {}, /*else*/ ...)`. This mirrors the established convention used by FIRRTL lowering and immediately fixes Quartus for everyone using `seq.fifo`.

Pros: tiny, follows precedent, no API change.
Cons: doesn't help other passes that emit `verif.*assert*` ops (e.g. anything user-written or future lowerings).

#### 2. Add an `ifdef`-guard option to `LowerVerifToSV`

Generalize the fix: give `VerifToSVPass` ([VerifToSV.cpp#L201-L226](https://github.com/llvm/circt/blob/main/lib/Conversion/VerifToSV/VerifToSV.cpp#L201-L226)) a pass option (e.g. `guard-with-ifndef=SYNTHESIS`, default off) that wraps every produced `sv.assert.property` / `sv.assume.property` / `sv.cover.property` (and the procedural variants) in an `sv::IfDefOp` with the configured macro name. Pipelines targeting FPGA synthesis can enable it from the driver.

Pros: covers all current and future producers of `verif.*` ops; opt-in so nothing changes by default.
Cons: slightly larger surface; needs a small amount of design (per-op vs. per-module wrap).

#### 3. Provide a small "strip / guard verification" utility pass

A standalone pass `--guard-verification-ops=SYNTHESIS` (or `--strip-verification-ops` for users who want them gone entirely) operating on `hw.module` bodies, wrapping/erasing all `verif::Assert*Op`, `verif::Assume*Op`, `verif::CoverOp`, and (post-`VerifToSV`) `sv::AssertPropertyOp` etc. This is composable and backend-agnostic; any pipeline can add it before ExportVerilog.

Pros: most flexible; gives users an explicit knob; useful even for non-FIFO assertions.
Cons: another pass to maintain; needs to know about both `verif` and `sv` flavors.

### Orthogonal: vendor-specific synthesis pragma support

Independently of the above, CIRCT could grow first-class support for emitting Quartus / Synopsys / Cadence-style line pragmas (e.g. `// synthesis translate_off` ... `// synthesis translate_on`) around assertions, in addition to (or instead of) `` `ifndef SYNTHESIS ``. This is what Quartus's error message literally suggests, and it doesn't require the user to define the `SYNTHESIS` macro in their tool flow.

This is orthogonal to the three fixes above: those decide *whether/where* a guard appears around verification ops; this decides *what kind of guard* is emitted. A natural place for this would be a new `LoweringOption` (e.g. `assertionGuardStyle=ifdef|translateOffOn|none`) consumed by ExportVerilog when emitting any `sv.assert*`/`sv.assume*`/`sv.cover*` op, regardless of how it got there. Because ExportVerilog is the consumer, `LoweringOptions` is appropriate for *this* knob.

## References

- `seq.fifo` assertion creation: [LowerSeqFIFO.cpp#L181-L201](https://github.com/llvm/circt/blob/main/lib/Dialect/Seq/Transforms/LowerSeqFIFO.cpp#L181-L201)
- `verif.clocked_assert` → `sv.assert.property` (no guard): [VerifToSV.cpp#L168-L201](https://github.com/llvm/circt/blob/main/lib/Conversion/VerifToSV/VerifToSV.cpp#L168-L201)
- `disallowClockedAssertions` option (insufficient for this issue): [LoweringOptions.h#L174-L178](https://github.com/llvm/circt/blob/main/include/circt/Support/LoweringOptions.h#L174-L178)
- Removal of `-sv-extract-test-code`: [Firtool.cpp#L814](https://github.com/llvm/circt/blob/main/lib/Firtool/Firtool.cpp#L814)
- Existing `addToIfDefBlock("SYNTHESIS", ...)` precedent in FIRRTL lowering: [LowerToHW.cpp#L5235](https://github.com/llvm/circt/blob/main/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L5235), [LowerToHW.cpp#L5283](https://github.com/llvm/circt/blob/main/lib/Conversion/FIRRTLToHW/LowerToHW.cpp#L5283), [FirRegLowering.cpp#L413](https://github.com/llvm/circt/blob/main/lib/Conversion/SeqToSV/FirRegLowering.cpp#L413)
- `LoweringOptions` consumers (only ExportVerilog and late SV passes): [ExportVerilog.cpp#L36](https://github.com/llvm/circt/blob/main/lib/Conversion/ExportVerilog/ExportVerilog.cpp#L36), [PrettifyVerilog.cpp#L553](https://github.com/llvm/circt/blob/main/lib/Dialect/SV/Transforms/PrettifyVerilog.cpp#L553), [HWLegalizeModules.cpp#L459](https://github.com/llvm/circt/blob/main/lib/Dialect/SV/Transforms/HWLegalizeModules.cpp#L459)

### PR DESCRIPTION
**Pull request by teqdruid**
Add `--sv-mask-non-synthesizable`, a late SV pass that hides concurrent and property assert/assume/cover ops from synthesis. Three modes (`delete`, `ifdef`) and a configurable macro name. All modes are idempotent across re-runs. Operates on `mlir::ModuleOp` and processes each `hw.module` in parallel. Wired into the PyCDE final lowering pipeline so its output is acceptable to Quartus and other tools that reject SVA constructs. This is the configurable-pass alternative to the always-on `ifdef SYNTHESIS` guards in `LowerVerifToSV` that were tried in #10387 and reverted.

Closes #10378.

Assisted-by: vscode:Claude Opus 4.7

### CODE REVIEW

**Review by uenoku — COMMENTED** [review]
Do we really need all three modes?

If `ifdef` is enough, could we just wrap each op with `ifdef` and let HWCleanup
merge guards where possible?

The merging condition here seems very narrow, since it only handles immediately
adjacent ops. For example, if the condition is created right before the
verification op, the guards cannot be merged. The implementation is also
non-trivial and overlaps with HWCleanup. HWCleanup seems to use a slightly better
heuristic: it can merge blocks even when there are non-adjacent ops between them,
as long as the ops in between are [pure](https://github.com/llvm/circt/blob/aafd14f20b7d3cc7cddc9e8f5ece4a64f7443f85/lib/Dialect/SV/Transforms/HWCleanup.cpp#L217-L229).

I understand that we cannot use HWCleanup for `sv.verbatim` / pragma mode. But
unless that mode is really needed, I wonder if `ifdef` + HWCleanup is enough.

I’m also a bit uncomfortable with checking the contents of `sv.verbatim` strings.
I believe the current implementation can generate odd output for a case like
this:

```
sv.assert
sv.verbatim "// synthesis translate_off"
sv.assert
sv.verbatim "// synthesis translate_on"
```
which could become something like:
```
sv.verbatim "// synthesis translate_off"
sv.assert
sv.verbatim "// synthesis translate_off"
sv.assert
sv.verbatim "// synthesis translate_on"
```

If we really want to enclose operations with `// synthesis translate_off/_on`, I think it's
justifiable to introduce a dedicated op for this, maybe something like `sv.nonsynthesizable`, since this is a very common pattern.

**Comment by teqdruid** [comment]
> Do we really need all three modes?
> 
> If `ifdef` is enough, could we just wrap each op with `ifdef` and let HWCleanup merge guards where possible?

Well... I prefer the pragma approach since tools later on use it automatically. No need to specify a macro. But given the amount of infrastructure we have to deal with ifdefs, that's a more robust decision.

> If we really want to enclose operations with `// synthesis translate_off/_on`, I think it's
> justifiable to introduce a dedicated op for this, maybe something like `sv.nonsynthesizable`, since this is a very common pattern.

Yeah, I agree. But unless we want to add support directly to ExportVerilog, it has to get lowered to verbatims at some point.

> I’m also a bit uncomfortable with checking the contents of `sv.verbatim` strings.
> I believe the current implementation can generate odd output for a case like
> this:

I share your concern. This was to ensure this pass was idempotent.

Given that most of the complexity comes from the pragma mode, I'll remove it and rely on hwcleanup to merge ifdef blocks.

**Comment by teqdruid** [comment]
@uenoku or @seldridge could you give this another look?

**Review by uenoku — APPROVED** [review]
LGTM

File: include/circt/Dialect/SV/SVPasses.td [hunk 1]
```diff
@@ -130,4 +130,44 @@ def HWEliminateInOutPorts : Pass<"hw-eliminate-inout-ports",
...
+def SVMaskNonSynthesizable : Pass<"sv-mask-non-synthesizable",
+                                  "mlir::ModuleOp"> {
+  let summary = "Hide non-synthesizable SV ops from synthesis tools";
+  let description = [{
+    This pass hides a fixed set of `sv` ops that are not generally
+    synthesizable (concurrent and property assertions) from the
+    SystemVerilog output. Two masking modes are supported:
+
+    * `delete` — erase the matched ops entirely.
+    * `ifdef`  — wrap each matched op individually in
```
Comment (seldridge) [inline include/circt/Dialect/SV/SVPasses.td]: Rage against the clanker em-dash, rage against the dying of the ASCII em-dash `---`:

```suggestion
    * `delete`: erase the matched ops entirely.
    * `ifdef`: wrap each matched op individually in
```

File: lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  Region *elseRegion = nullptr;
+  if (auto ifdef = dyn_cast<sv::IfDefOp>(op)) {
+    macro = ifdef.getCond().getName();
+    elseRegion = &ifdef.getElseRegion();
+  } else if (auto ifdef = dyn_cast<sv::IfDefProceduralOp>(op)) {
+    macro = ifdef.getCond().getName();
+    elseRegion = &ifdef.getElseRegion();
+  } else {
+    return nullptr;
+  }
```
Comment (seldridge) [inline lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp]: Nit: a TypeSwitch would allow the PR to collapse the two identical if/else if blocks via templates.

File: lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+/// the procedural variant based on the enclosing region.
+static void maskOpByIfdef(Operation *op, StringRef macro) {
+  bool procedural = isInProceduralRegion(op);
+  OpBuilder builder(op);
+  Location loc = op->getLoc();
+
+  // Passing a non-null `elseCtor` is what triggers the builder to create an
+  // else block.
+  Block *elseBlock;
+  if (procedural)
```
Comment (seldridge) [inline lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp]: Nit: single use variable defined much earlier is more readable inline:


```suggestion
  if (isInProceduralRegion(op))
```

File: lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+      maskOpByIfdef(&op, macro);
+      changed = true;
+      continue;
+    }
+    Region *guardedElseRegion = getMatchingIfDefElseRegion(&op, macro);
+    for (Region &region : op.getRegions()) {
+      if (&region == guardedElseRegion)
+        continue;
+      for (Block &nested : region)
+        changed |= processBlockIfdef(nested, macro);
```
Comment (seldridge) [inline lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp:140]: Recursion over blocks is allowed. 👍

File: lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    switch (mode) {
+    case MaskNonSynthesizableMode::Delete:
+      processBlockDelete(body);
+      break;
+    case MaskNonSynthesizableMode::Ifdef:
+      if (processBlockIfdef(body, passDownMacro))
+        anyChanged.store(true, std::memory_order_relaxed);
+      break;
+    }
+  });
```
Comment (seldridge) [inline lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp]: `parallelTransformReduce` may be slightly better in highly parallel environments by avoiding the lock.

Also, usually an LLVM atomic is more expected than an std one.

File: test/Dialect/SV/sv-mask-non-synthesizable.mlir [hunk 6]
```diff
@@ -0,0 +1,222 @@
+// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=delete %s | FileCheck %s --check-prefix=DELETE
+// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=ifdef %s | FileCheck %s --check-prefix=IFDEF
+// Default mode is `ifdef`.
+// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable %s | FileCheck %s --check-prefix=IFDEF
+// Both modes are idempotent across re-runs.
+// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=delete %s | circt-opt -split-input-file -sv-mask-non-synthesizable=mode=delete | FileCheck %s --check-prefix=DELETE
+// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=ifdef %s | circt-opt -split-input-file -sv-mask-non-synthesizable=mode=ifdef | FileCheck %s --check-prefix=IFDEF
+// `ifdef` mode honors a custom macro name and is idempotent under it.
+// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable='mode=ifdef macro=MY_GUARD' %s | FileCheck %s --check-prefix=CUSTOM
+// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable='mode=ifdef macro=MY_GUARD' %s | circt-opt -split-input-file -sv-mask-non-synthesizable='mode=ifdef macro=MY_GUARD' | FileCheck %s --check-prefix=CUSTOM
```
Comment (seldridge) [inline test/Dialect/SV/sv-mask-non-synthesizable.mlir]: Mega-nit: when there are multiple run lines like this that are testing distinct things, I find that adding newlines helps with readability:


```suggestion
// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=delete %s | FileCheck %s --check-prefix=DELETE
// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=ifdef %s | FileCheck %s --check-prefix=IFDEF
//
// Default mode is `ifdef`.
// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable %s | FileCheck %s --check-prefix=IFDEF
//
// Both modes are idempotent across re-runs.
// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=delete %s | circt-opt -split-input-file -sv-mask-non-synthesizable=mode=delete | FileCheck %s --check-prefix=DELETE
// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable=mode=ifdef %s | circt-opt -split-input-file -sv-mask-non-synthesizable=mode=ifdef | FileCheck %s --check-prefix=IFDEF
// 
// `ifdef` mode honors a custom macro name and is idempotent under it.
// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable='mode=ifdef macro=MY_GUARD' %s | FileCheck %s --check-prefix=CUSTOM
// RUN: circt-opt -split-input-file -sv-mask-non-synthesizable='mode=ifdef macro=MY_GUARD' %s | circt-opt -split-input-file -sv-mask-non-synthesizable='mode=ifdef macro=MY_GUARD' | FileCheck %s --check-prefix=CUSTOM
```

This is generally very good and thorough. 👍

Comment (seldridge) [reply test/Dialect/SV/sv-mask-non-synthesizable.mlir]: The idempotency check of the custom ifdef guard seems unnecessary.

File: test/Dialect/SV/sv-mask-non-synthesizable.mlir [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+// CUSTOM: sv.macro.decl @MY_GUARD
+
+// DELETE-LABEL: hw.module @basic
+// IFDEF-LABEL:  hw.module @basic
+// CUSTOM-LABEL: hw.module @basic
+hw.module @basic(in %clock : i1, in %prop : i1) {
+  // Untouched ops outside the masked set always pass through.
+  // DELETE: hw.constant true
+  // IFDEF:  hw.constant true
+  // CUSTOM: hw.constant true
```
Comment (seldridge) [inline test/Dialect/SV/sv-mask-non-synthesizable.mlir]: This is unnecessary. I'd probably just drop it, but it could use a common prefix for the common things.

File: test/Dialect/SV/sv-mask-non-synthesizable.mlir [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+// Each masked op is wrapped in its own `sv.ifdef`; adjacent wrappers are not
+// coalesced (we rely on `hw-cleanup` to merge them downstream).
+//===----------------------------------------------------------------------===//
+
+// IFDEF: sv.macro.decl @SYNTHESIS
+// CUSTOM: sv.macro.decl @MY_GUARD
+
+// DELETE-LABEL: hw.module @basic
+// IFDEF-LABEL:  hw.module @basic
+// CUSTOM-LABEL: hw.module @basic
```
Comment (seldridge) [inline test/Dialect/SV/sv-mask-non-synthesizable.mlir]: Consider using a common label to collapse these.

File: test/Dialect/SV/sv-mask-non-synthesizable.mlir [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+hw.module @SYNTHESIS() {}
+
+// IFDEF-LABEL: hw.module @collides_with_module
+hw.module @collides_with_module(in %clock : i1, in %prop : i1) {
+  // IFDEF:      sv.ifdef @SYNTHESIS_0 {
+  // IFDEF-NEXT: } else {
+  // IFDEF-NEXT:   sv.assert.concurrent posedge %clock, %prop
+  // IFDEF-NEXT: }
+  sv.assert.concurrent posedge %clock, %prop
+}
```
Comment (seldridge) [inline test/Dialect/SV/sv-mask-non-synthesizable.mlir:180]: Nice test.

File: test/Dialect/SV/sv-mask-non-synthesizable.mlir [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+sv.macro.decl @MY_SYM["SYNTHESIS"]
+
+// IFDEF-LABEL: hw.module @reuses_existing_decl
+hw.module @reuses_existing_decl(in %clock : i1, in %prop : i1) {
+  // IFDEF:      sv.ifdef @MY_SYM {
+  // IFDEF-NEXT: } else {
+  // IFDEF-NEXT:   sv.assert.concurrent posedge %clock, %prop
+  // IFDEF-NEXT: }
+  sv.assert.concurrent posedge %clock, %prop
+}
```
Comment (seldridge) [inline test/Dialect/SV/sv-mask-non-synthesizable.mlir:200]: Very nice edge case test.

**Review by seldridge — APPROVED** [review]
Only to aggressively conserve concepts, would it be better to _not_ have the delete option and instead rely on a Verilog preprocessor in a build flow? I'm not against having the option, though! Just a thought.

This looks great. Thanks for the iteration, John.

**Comment by teqdruid** [comment]
> Only to aggressively conserve concepts, would it be better to _not_ have the delete option and instead rely on a Verilog preprocessor in a build flow? I'm not against having the option, though! Just a thought.

It's for cases where users cannot or don't want to change macros in the build.

File: lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp (hunk 2 above)
Comment (teqdruid) [reply lib/Dialect/SV/Transforms/SVMaskNonSynthesizable.cpp]: I had no idea it did that!
